packages/create-subscription/src/createSubscription.js

```diff
@@ -0,0 +1,153 @@
+  |}>,
+): React$ComponentType<{
+  children: (value: Value | void) => React$Node,
+  source: any,
+}> {
+  const {getValue, subscribe} = config;
+
+  warning(
+    typeof getValue === 'function',
+    'Subscription must specify a getValue function',
+  );
+  warning(
+    typeof subscribe === 'function',
+    'Subscription must specify a subscribe function',
+  );
+
+  type Props = {
+    children: (value: Value) => React$Element<any>,
+    source: any,
+  };
+  type State = {
+    source: Property,
+    unsubscribeContainer: {
+      unsubscribe?: Unsubscribe | CannotUnsubscribe,
+    },
+    value: Value | void,
+  };
+
+  // Reference: https://gist.github.com/bvaughn/d569177d70b50b58bff69c3c4a5353f3
+  class Subscription extends React.Component<Props, State> {
+    state: State = {
+      source: this.props.source,
+      unsubscribeContainer: {},
+      value:
+        this.props.source != null ? getValue(this.props.source) : undefined,
+    };
+
+    static getDerivedStateFromProps(nextProps, prevState) {
+      if (nextProps.source !== prevState.source) {
+        return {
+          source: nextProps.source,
+          unsubscribeContainer: {},
+          value:
+            nextProps.source != null ? getValue(nextProps.source) : undefined,
+        };
+      }
+
+      return null;
+    }
+
+    componentDidMount() {
+      this.subscribe();
+    }
+
+    componentDidUpdate(prevProps, prevState) {
+      if (this.state.source !== prevState.source) {
+        this.unsubscribe(prevState);
+        this.subscribe();
+      }
+    }
+
+    componentWillUnmount() {
+      this.unsubscribe(this.state);
+    }
+
+    render() {
+      return this.props.children(this.state.value);
+    }
+
+    subscribe() {
+      const {source} = this.state;
+      if (source != null) {
+        const callback = (value: Value | void) => {
+          this.setState(state => {
+            // If the value is the same, skip the unnecessary state update.
+            if (value === state.value) {
+              return null;
+            }
+
+            // If this event belongs to an old or uncommitted data source, ignore it.
+            if (source !== state.source) {
+              return null;
+            }
+
+            return {value};
+          });
+        };
+
+        // Store subscription for later (in case it's needed to unsubscribe).
+        // This is safe to do via mutation since:
+        // 1) It does not impact render.
+        // 2) This method will only be called during the "commit" phase.
+        const unsubscribeOrBoolean = subscribe(source, callback);
+
+        invariant(
+          unsubscribeOrBoolean === false ||
+            typeof unsubscribeOrBoolean === 'function',
+          'A subscription should return either an unsubscribe function or false.',
+        );
+
+        this.state.unsubscribeContainer.unsubscribe = unsubscribeOrBoolean;
+
+        // External values could change between render and mount,
```

Comment:
Maybe move this *above* the part where you make the subscription? That way, in the rare case that a subscription resolves synchronously the first time, you get this behavior:

1. `getValue` which might potentially be related to a stale/clowny implementation like the one I mentioned above.
2. Perform the subscription, which might resolve sync and clobber the value in step 1.

Reply:
I don't see how moving this line would affect the case you mentioned above?

Reply:
I don't think I want to _try_ to support the clowny pattern you mentioned above. If the "subscribable" doesn't invoke callbacks on-subscribe, then that pattern will definitely break things. Even if it does, but doesn't sync invoke them, it will cause unnecessary double renders. I'd rather fail "hard" in this case.